fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/VectorIndexMaintainer.java

```diff
@@ -0,0 +1,493 @@
+    }
+
+    @SuppressWarnings({"resource", "PMD.CloseResource"})
+    private CompletableFuture<Optional<Tuple>> nextPrefixTuple(@Nonnull final TupleRange prefixRange,
+                                                               final int prefixSize,
+                                                               @Nullable final Tuple lastPrefixTuple,
+                                                               @Nonnull final ScanProperties scanProperties) {
+        final Subspace indexSubspace = getIndexSubspace();
+        final KeyValueCursor cursor;
+        if (lastPrefixTuple == null) {
+            cursor = KeyValueCursor.Builder.withSubspace(indexSubspace)
+                    .setContext(state.context)
+                    .setRange(prefixRange)
+                    .setContinuation(null)
+                    .setScanProperties(scanProperties.setStreamingMode(CursorStreamingMode.ITERATOR)
+                            .with(innerExecuteProperties -> innerExecuteProperties.setReturnedRowLimit(1)))
+                    .build();
+        } else {
+            KeyValueCursor.Builder builder = KeyValueCursor.Builder.withSubspace(indexSubspace)
+                    .setContext(state.context)
+                    .setContinuation(null)
+                    .setScanProperties(scanProperties)
+                    .setScanProperties(scanProperties.setStreamingMode(CursorStreamingMode.ITERATOR)
+                            .with(innerExecuteProperties -> innerExecuteProperties.setReturnedRowLimit(1)));
+
+            cursor = builder.setLow(indexSubspace.pack(lastPrefixTuple), EndpointType.RANGE_EXCLUSIVE)
+                    .setHigh(prefixRange.getHigh(), prefixRange.getHighEndpoint())
+                    .build();
+        }
+
+        return cursor.onNext().thenApply(next -> {
+            cursor.close();
+            if (next.hasNext()) {
+                final KeyValue kv = Objects.requireNonNull(next.get());
+                return Optional.of(TupleHelpers.subTuple(indexSubspace.unpack(kv.getKey()), 0, prefixSize));
+            }
+            return Optional.empty();
+        });
+    }
+
+    @Override
+    protected <M extends Message> CompletableFuture<Void> updateIndexKeys(@Nonnull final FDBIndexableRecord<M> savedRecord,
+                                                                          final boolean remove,
+                                                                          @Nonnull final List<IndexEntry> indexEntries) {
+        final KeyWithValueExpression keyWithValueExpression = getKeyWithValueExpression(state.index.getRootExpression());
+        final int prefixSize = keyWithValueExpression.getColumnSize();
+        final Subspace indexSubspace = getIndexSubspace();
+        final var futures = indexEntries.stream().map(indexEntry -> {
+            final var indexKeyItems = indexEntry.getKey().getItems();
+            final Tuple prefixKey = Tuple.fromList(indexKeyItems.subList(0, prefixSize));
```

Comment:
If this is a key with values, should the index key items always have `prefixSize` items? Is there a reason this isn't just `indexEntry.getKey()`?

There is, granted, some amount of ambiguity about what actually goes in the index key. Is the concern that it would include the primary key?